### PR TITLE
C4-133 Check result of ES deletion

### PR DESCRIPTION
**Pull request by willronchetti**
- Previously, we would blindly delete from the DB regardless of whether the ES deletion was successful or not.
- We no longer do that now depending on the scenario. An internal server error will be thrown if a deletion is unsuccessful in ES, and the purge from the DB is aborted.

### CODE REVIEW

File: snovault/elasticsearch/esstorage.py [hunk 1]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
-        index_name = get_namespaced_index(self.registry, item_type)
+    @staticmethod
+    def delete_from_es(es, rid, index_name, item_type):
+        """ Helper method that deletes an item from Elasticsearch """
         try:
-            self.es.delete(id=rid, index=index_name, doc_type=item_type)
+            es.delete(id=rid, index=index_name, doc_type=item_type)
         except elasticsearch.exceptions.NotFoundError:
             # Case: Not yet indexed
             log.error('PURGE: Could not find %s in ElasticSearch. Continuing.' % rid)
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:241]: I think either returning after the log call or at least setting `result = True` or `result = False` and then doing `return result` would be better. As it is, it's hard to tell if you intend to `return True` here or if it just accidentally falls through.

File: snovault/elasticsearch/esstorage.py [hunk 2]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
-    def purge_uuid(self, rid, item_type, max_sid=None):
-        """
-        Purge a uuid from the write storage (Elasticsearch)
-        If there is a mirror environment set up for the indexer, also attempt
-        to remove the uuid from the mirror Elasticsearch
-        """
-        index_name = get_namespaced_index(self.registry, item_type)
+    @staticmethod
+    def delete_from_es(es, rid, index_name, item_type):
+        """ Helper method that deletes an item from Elasticsearch """
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:236]: Should document the meaning of the return value. Without that, I can't tell which clauses are returning the right or wrong value.

File: snovault/elasticsearch/esstorage.py [hunk 3]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
             log.error('PURGE: Cannot delete %s in ElasticSearch. Error: %s Continuing.' % (item_type, str(exc)))
+            return False
         else:
             log.info('PURGE: successfully deleted %s in ElasticSearch' % rid)
+        return True
+
+    def purge_uuid_from_mirror(self, rid, item_type):
+        """ Helper method for purge_uuid that purges rid from the mirror on ESStorage """
+        mirror_env = self.registry.settings['mirror.env.name']
+        mirror_health = self.registry.settings['mirror_health']
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:252]: What's in this variable? I think this is the result of the previous health report, so maybe call it `last_health_report` or `cached_health_check_response` or something?

File: snovault/elasticsearch/esstorage.py [hunk 4]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
+        use_aws_auth = self.registry.settings.get('elasticsearch.aws_auth')
+        log.info('PURGE: attempting to purge %s from mirror storage %s' % (rid, mirror_env))
+
+        # if we could not get mirror health, bail here
+        if 'error' in mirror_health:
+            log.error(
+                'PURGE: Tried to purge %s from mirror storage but couldn\'t get health page. Is staging up?' % rid)
+            log.error('PURGE: Mirror health error: %s' % mirror_health)
+            log.error('PURGE: Recomputing mirror health...')
+            self.registry.settings['mirror_health'] = ff_utils.get_health_page(ff_env=mirror_env)
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:262]: What if this returns an error? Why are we computing it at all here, slowing us down. If we don't care about the value, why not spin off a thread to do it and not wait? What happens if it is repeatedly down?

Comment (netsettler) [reply snovault/elasticsearch/esstorage.py:262]: I'm having real trouble untangling what this is doing? Can you make a comment ahead of all this that says what the relevant state is and how it's intended to be maintained so I can check this implementation against the plan? I feel like there are a lot of working parts with subtle effects and a big chance to do things wrong.

File: snovault/elasticsearch/esstorage.py [hunk 5]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
+            return True  # in this case proceed with the DB deletion
+
+        # make sure use_aws_auth is bool
+        if not isinstance(use_aws_auth, bool):
+            use_aws_auth = True if use_aws_auth == 'true' else False
+
+        if not self.mirror_client:  # cached on this object
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:269]: I think the comment is telling me something but I'm not sure what. Are there multiple objects to choose from?

File: snovault/elasticsearch/esstorage.py [hunk 6]
```diff
@@ -231,57 +231,69 @@ def get_max_sid(self):
...
+            self.mirror_client = es_utils.create_es_client(mirror_health['elasticsearch'],
+                                                            use_aws_auth=use_aws_auth)
+
+        mirror_index = mirror_health.get('namespace', '') + item_type
```
Comment (netsettler) [inline snovault/elasticsearch/esstorage.py:273]: This 'addition' on the fly seems error-prone. Can you separate this out into something organized. I don't like seeing a low-level operation like this in a high-level operation. Better to do something that isolates the behavior and can test what happens when the namespace is missing, etc.

We're going up and down in the abstraction hierarchy here in a way that's hard to feel is good. If possible, build up functions into a vocabulary that all speak at a similar level of abstraction and then do the logic at that level, peeling back to implementation details within each of those operations so those lower level operations can be separately found, tested, etc.

File: snovault/elasticsearch/esstorage.py (hunk 4 above)
Comment (willronchetti) [reply snovault/elasticsearch/esstorage.py:262]: This is a bug that will be fixed in the next update. We do care about this value and will need it.

File: snovault/elasticsearch/esstorage.py (hunk 5 above)
Comment (willronchetti) [reply snovault/elasticsearch/esstorage.py:269]: If we've done this before using this object, we do not need to regenerate the mirror client.

File: snovault/elasticsearch/esstorage.py (hunk 6 above)
Comment (willronchetti) [reply snovault/elasticsearch/esstorage.py]: This is done all over the place and is required. Yes, it is based on an assumption but it is an assumption we are careful not to break. The health page should always have the namespace on it (in addition to existing in the application registry), and the index is always built with namespace + item_type. I've created a helper method in `indexer_utils` that does this operation, like `get_namespaced_index`.

**Pull request merged**